rosplan_planning_system/src/PlanDispatch/EsterelPlanDispatcher.cpp

```diff
@@ -3,396 +3,486 @@
+                        float maximum_dispatch_time = node_real_dispatch_time[edge.source_ids[0]] + edge.duration_upper_bound - planStartTime;
+
+                        // widen bounds based on parameter (default by 0)
+                        minimum_dispatch_time = minimum_dispatch_time - action_timeout_fraction*minimum_dispatch_time;
+                        maximum_dispatch_time = maximum_dispatch_time + action_timeout_fraction*maximum_dispatch_time;
+
+                        // check the current time with the lower bound
+                         double NOW = ros::Time::now().toSec();
+                         if (NOW - planStartTime < minimum_dispatch_time) { 
+                            /*if(node.node_type == rosplan_dispatch_msgs::EsterelPlanNode::ACTION_END && action_completed[node.action.action_id]) {
+                                ROS_INFO("KCL: (%s) Action completed too early, %s: %f < %f.", ros::this_node::getName().c_str(), node.name.c_str(), NOW-planStartTime, minimum_dispatch_time);
+                                replan_requested = true;
+                            }*/
+                            times_activate_action = false;
+                            finished_execution = false;
+                            break;
+                        }
+
+                        // check the current time with the upper bound
+                        if (NOW - planStartTime > maximum_dispatch_time) {
+                            // don't check deadlines for actions that have actually completed.
+                            if(node.node_type != rosplan_dispatch_msgs::EsterelPlanNode::ACTION_END || !action_completed[node.action.action_id]) {
+                                replan_requested =  true;
+                                times_activate_action = false;
+                                ROS_INFO("KCL: (%s) Deadline passed, %s: %f > %f.", ros::this_node::getName().c_str(), node.name.c_str(), NOW-planStartTime, maximum_dispatch_time);
+                            }
+                        }
+                    }
+
+                    // the lower bound of the action is not yet reached
+                    if(!times_activate_action) continue;
+                }
+
+                // dispatch new action
+                if(node.node_type == rosplan_dispatch_msgs::EsterelPlanNode::ACTION_START && !action_dispatched[node.action.action_id]) {
+
+                    finished_execution = false;
+
+                    // query KMS for condition edges
+                    bool condition_activate_action = false;
+                    if(edges_activate_action) {
+                        condition_activate_action = checkPreconditions(node.action);
+                    }
+
+                    // the state is unexpected
+                    if(!condition_activate_action) {
+                        replan_requested = true;
```

Comment:
This might cause some unwanted side-effects. In existing systems without temporal constraints, action are often simply waiting for preconditions to become true before executing. It is undesirable to replan instead of waiting.

This IF statement should also include the "timeout_actions" flag.